Assets/00_Content/Scripts/Vikings/Viking.cs

```diff
@@ -112,5 +169,76 @@ public class Viking : Interactable {
 		public override bool CanInteract(GameObject player, PickUp item) {
 			return state.CanInteract(player, item);
 		}
+
+		private void OnTriggerEnter(Collider other) {
+			if (weaponLayer.ContainsLayer(other.gameObject.layer)) {
+				RegisterHitFromPlayer(other);
+			}
+		}
+
+		private void RegisterHitFromPlayer(Collider other) {
+
+			if (state is LeavingVikingState)
+				return;
+
+			//TODO - Probably get some more generic type of weapon instead of axe. Make an interface for all weapons?
+			Axe axe = other.gameObject.GetComponent<Axe>();
+
+			if (axe.IsAttacking && !isAttacked) {
+
+				isAttacked = true;
+
+				foreach (MeshRenderer hitHighlightMeshRenderer in hitHighlightMeshRenderers) {
+					hitHighlightMeshRenderer.material = brawlingMaterial;
+				}
+
+				if (!IsSeated) {
+					PlayerComponent playerComponent = axe.GetComponentInParent<PlayerComponent>();
+					Vector3 direction = (playerComponent.transform.position - transform.position).normalized;
+					navMeshAgent.enabled = false;
+					rb.isKinematic = false;
+					rb.AddForce(direction * axe.WeaponData.knockBackStrength * -1, ForceMode.Impulse);
+				}
+
+				StartCoroutine(ResetHitSimulation());
+
+				if (state is BrawlingVikingState) {
+					Stats.TakeBrawlDamage(axe.WeaponData.brawlDamage);
+					if (Stats.BrawlHealth <= 0) {
+
+						ChangeState(new LeavingVikingState(this));
+						return;
+					}
+				}
+
+				if (state is WaitingForSeatVikingState) {
+					ChangeState(new LeavingVikingState(this));
+					return;
+				}
+
+				Stats.TakeMoodDamage(axe.WeaponData.moodDamage);
+
+				if (Stats.Mood <= 15f) {
+					ChangeState(new BrawlingVikingState(this, axe.GetComponentInParent<PlayerComponent>()));
+				}
+			}
+		}
+
+		private IEnumerator ResetHitSimulation() {
+
+			yield return new WaitForSeconds(0.5f);
+
+			foreach (MeshRenderer hitHighlightMeshRenderer in hitHighlightMeshRenderers) {
+				hitHighlightMeshRenderer.material = normalMaterial;
+			}
+
+			if (!IsSeated) {
+				rb.isKinematic = true;
+				navMeshAgent.enabled = true;
+				state.Enter();
```

Comment:
It seems like a really bad idea to call enter on a state directly. This assumes that all the internal state of the state will get reset properly, but can that really be guaranteed?

Reply:
Nothing seems to get lost in the process, all of the states that get affected by this is will reset enough. I´m doing this as "SetDestination" has to be set again on the currentState after disabling/renenabling the navMesh and this is easiest trough the Enter method

Reply:
Yes, but my point is that the Enter() function was ever only supposed to be called once in the lifetime of a state. When writing a state you should be able to make that assumtion. Even though it works fine with all states now, it might be a source of bugs in the future.